driver-core/src/test/functional/com/mongodb/client/CrudTestHelper.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client;
+
+import org.bson.BsonArray;
+import org.bson.BsonDocument;
+import org.bson.BsonType;
+import org.bson.BsonValue;
+
+import static org.junit.Assert.assertEquals;
+
+public final class CrudTestHelper {
+
+    public static void replaceTypeAssertionWithActual(final BsonDocument expected, final BsonDocument actual) {
```

Comment:
Supports the $$type syntax required by the test runner.  It recursively looks for $$type key and replaces the value with the actual value, or asserts if the actual value is the wrong type.